java-frontend/src/main/java/org/sonar/java/se/checks/NoWayOutLoopCheck.java

```diff
@@ -0,0 +1,224 @@
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.UnaryExpressionTree;
+import org.sonar.plugins.java.api.tree.WhileStatementTree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+@Rule(
+  key = "S2189",
+  name = "Loops should not be infinite",
+  priority = Priority.BLOCKER,
+  tags = {"bug", "misra"})
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("15min")
+public class NoWayOutLoopCheck extends SECheck {
+
+  private enum UpdateTypes {
+    INCREMENT, DECREMENT, INDETERMINATE
+  }
+
+  Map<Tree, CFGLoop> loopStarts;
+
+  @Override
+  public void init(MethodTree tree, CFG cfg) {
+    loopStarts = CFGLoop.getCFGLoops(cfg);
+  }
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    final PreStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  private class PreStatementVisitor extends CheckerTreeNodeVisitor {
+
+    private final CheckerContext context;
+
+    protected PreStatementVisitor(CheckerContext context) {
+      super(context.getState());
+      this.context = context;
+    }
+
+    @Override
+    public void visitWhileStatement(WhileStatementTree tree) {
+      if (CFGLoop.isHardCodedTrue(tree.condition())) {
+        CFGLoop loopBlocks = loopStarts.get(tree);
+        if (loopBlocks.hasNoWayOut()) {
+          context.reportIssue(tree, NoWayOutLoopCheck.this, "Add an end condition to this loop.");
+        }
+      }
+    }
+
+    @Override
+    public void visitForStatement(ForStatementTree tree) {
+      CFGLoop loopBlocks = loopStarts.get(tree);
+      if (tree.condition() == null) {
+        if (loopBlocks.hasNoWayOut()) {
+          context.reportIssue(tree, NoWayOutLoopCheck.this, "Add an end condition to this loop.");
+        }
+      } else if (isConditionUnreachable(tree)) {
+        context.reportIssue(tree, NoWayOutLoopCheck.this, "Correct this loop's end condition.");
+      }
+    }
+
+    private boolean isConditionUnreachable(ForStatementTree tree) {
+      UpdatesCollector collector = new UpdatesCollector();
+      tree.accept(collector);
+      ConditionType condition = new ConditionType(tree.condition(), collector);
+      return !condition.isMatched();
+    }
+  }
+
+  private static class UpdateType {
+
+    private Symbol symbol = null;
+    private UpdateTypes type = null;
+
+    UpdateType(Symbol symbol, UpdateTypes type) {
+      this.symbol = symbol;
+      this.type = type;
+    }
+
+    UpdateTypes type() {
+      return type;
+    }
+
+    boolean concerns(ExpressionTree operand) {
+      if (operand.is(Tree.Kind.IDENTIFIER)) {
+        return symbol == ((IdentifierTree) operand).symbol();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)